src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -1684,6 +1785,92 @@ UndistributeTable(Oid relationId)
 	{
 		ereport(ERROR, (errmsg("could not finish SPI connection")));
 	}
+	
+	if (cascadeToColocated)
+	{
+		Oid colocatedTableId = InvalidOid;
+		// For now we only support cascade to colocation for alter_distributed_table UDF
+		Assert(conversionType == ALTER_DISTRIBUTED_TABLE);
+		foreach_oid(colocatedTableId, colocatedTableList)
+		{
+			if (colocatedTableId == relationId)
+			{
+				continue;
+			}
+			if (conversionType == ALTER_DISTRIBUTED_TABLE)
+			{
+				AlterDistributedTable(colocatedTableId, NULL, shardCount, relationName, false);
+			}
+		}
+	}
+}
+
+
+/*
+ * UndistributeTable undistributes the given table. It uses ConvertTable function to
+ * create a new local table and move everything to that table.
+ * 
+ * The local tables, tables with references, partition tables and foreign tables are
+ * not supported. The function gives errors in these cases.
+ */
+void
+UndistributeTable(Oid relationId)
+{
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("cannot undistribute table"),
+						errdetail("because no such distributed table exists")));
+	}
+
+	relation_close(relation, NoLock);
+
+	if (!IsCitusTable(relationId))
+	{
+		ereport(ERROR, (errmsg("cannot undistribute table "),
+						errdetail("because the table is not distributed")));
+	}
+
+	EnsureTableNotReferencing(relationId);
+	EnsureTableNotReferenced(relationId);
+	EnsureTableNotForeign(relationId);
+	EnsureTableNotPartition(relationId);
+
+	ConvertTable(UNDISTRIBUTE_TABLE, relationId, NULL, 0, NULL, false);
+}
+
+
+/*
+ * AlterDistributedTable changes some properties of the given table. It uses
+ * ConvertTable function to create a new local table and move everything to that table.
+ * 
+ * The local tables, tables with references, partition tables and foreign tables are
+ * not supported. The function gives errors in these cases.
+ */
+void
+AlterDistributedTable(Oid relationId, char *distributionColumn, int shardCount, char *colocateWithTableName, bool cascadeToColocated)
+{
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("cannot undistribute table"),
+						errdetail("because no such distributed table exists")));
+	}
+	relation_close(relation, NoLock);
+
+	if (!IsCitusTable(relationId))
```

Comment:
could use IsCitusTableType here to check whether it's actually distributed (not reference)